File: src/it/scala/com/exasol/cloudetl/kafka/docker/KafkaImportIT.scala
package com.exasol.cloudetl.kafka

import java.sql.ResultSet
import java.util.UUID

import com.exasol.dbbuilder.dialects.Table
import com.exasol.matcher.ResultSetStructureMatcher.table
import com.exasol.matcher.TypeMatchMode.NO_JAVA_TYPE_CHECK

import org.hamcrest.Matcher
import org.hamcrest.MatcherAssert.assertThat
import org.scalatest.BeforeAndAfterEach

class KafkaImportIT extends BaseKafkaDockerIntegrationTest with BeforeAndAfterEach {

  private[this] val schemaName = "KAFKA_SCHEMA"
  private[this] val tableName = "KAFKA_TABLE"
  private[this] var topicName: String = _

  private[this] def getTableName(): String = s""""$schemaName"."$tableName""""

  override final def beforeAll(): Unit = {
    super.beforeAll()
    installKafkaConnector(schemaName)
    updateExasolHostsFile()
  }

  override final def beforeEach(): Unit = {
    topicName = "kafka-topic-" + UUID.randomUUID()
    createTopic(topicName)
  }

  override final def afterEach(): Unit = {
    executeStmt(s"DROP TABLE IF EXISTS ${getTableName()}")
    deleteTopic(topicName)
  }

  test("import string values") {
    KafkaImportChecker(Map("COLUMN" -> "VARCHAR(20)"))
      .withTopicValues(Seq("a", "b", "c"))
      .assert(
        table()
          .row("a", 0L, 0L)
          .row("b", 0L, 1L)
          .row("c", 0L, 2L)
          .matches(NO_JAVA_TYPE_CHECK)
      )
  }

  test("import with CONSUME_ALL_OFFSET parameter empty topic") {
    KafkaImportChecker(Map("COLUMN" -> "VARCHAR(20)"))
      .withImportProperties(Map("CONSUME_ALL_OFFSETS" -> "true"))
      .assert(table("VARCHAR", "BIGINT", "DECIMAL").matches())
  }

  test("import with CONSUME_ALL_OFFSET parameter") {
    val properties = Map(
      "CONSUME_ALL_OFFSETS" -> "true",
      "MAX_POLL_RECORDS" -> "2",
      "MIN_RECORDS_PER_RUN" -> "1",
      "MAX_RECORDS_PER_RUN" -> "2"
    )
    KafkaImportChecker(Map("COLUMN" -> "VARCHAR(20)"))
      .withImportProperties(properties)
      .withTopicValues(Seq("one", "two", "three"))
      .assert(
        table()
          .row("one", 0L, 0L)
          .row("two", 0L, 1L)
          .row("three", 0L, 2L)
          .matches(NO_JAVA_TYPE_CHECK)
      )
  }

  def updateExasolHostsFile(): Unit = {
    val brokerIpAddress = getContainerNetworkAddress(kafkaBrokerContainer)
    val schemaRegistryIpAddress = getContainerNetworkAddress(schemaRegistryContainer)
    val commands = Seq(
      s"sed -i '/kafka01/d' /etc/hosts",
      s"sed -i '/schema-registry/d' /etc/hosts",
      s"echo '$brokerIpAddress kafka01' >> /etc/hosts",
      s"echo '$schemaRegistryIpAddress schema-registry' >> /etc/hosts"
    )
    commands.foreach { case cmd =>
      val exitCode = exasolContainer.execInContainer("/bin/sh", "-c", cmd)
      if (exitCode.getExitCode() != 0) {
        throw new RuntimeException(s"Could not run command '$cmd' in Exasol container.")
      }
    }
  }

  case class KafkaImportChecker(columns: Map[String, String] = Map.empty) {
    private[this] val table = getTable()
    private[this] var properties: Map[String, String] = _

    def withImportProperties(props: Map[String, String]): KafkaImportChecker = {
      properties = props
      this
    }

    def withTopicValues(values: Seq[String]): KafkaImportChecker = {
      produceRecords(topicName, values)
      this
    }

    def assert(matcher: Matcher[ResultSet]): Unit = {
      assertResultSet(assertThat(_, matcher))
      ()
    }

    private[this] def getTable(): Table = {
      var tableBuilder = schema.createTableBuilder(tableName)
      val metadataColumns = Map(
        "KAFKA_PARTITION" -> "DECIMAL(18, 0)",
        "KAFKA_OFFSET" -> "DECIMAL(36, 0)"
      )
      (columns ++ metadataColumns).foreach { case (columnName, columnType) =>
        tableBuilder = tableBuilder.column(columnName, columnType)
      }
      tableBuilder.build()
    }

    private[this] def getImportStatement(): String = {
      val builder = new StringBuilder(defaultImportStatement(table))
      if (properties == null || !properties.contains("RECORD_KEY_FORMAT")) {
        builder.append("\n").append("RECORD_KEY_FORMAT = 'string'")
      }
      if (properties == null || !properties.contains("RECORD_VALUE_FORMAT")) {
        builder.append("\n").append("RECORD_VALUE_FORMAT = 'string'")
      }
      if (properties != null) {
        properties.foreach { case (key, value) =>
          builder.append("\n").append(s"$key = '$value'")
        }
      }
      builder.append(";")
      builder.toString()
    }

    private[this] def assertResultSet(block: ResultSet => Unit): Unit = {
      executeStmt(getImportStatement())
      val rs = executeQuery(s"SELECT * FROM ${getTableName()}")
      block(rs)
      rs.close()
    }
  }

  private[this] def defaultImportStatement(table: Table): String =
    s"""|IMPORT INTO ${table.getFullyQualifiedName()}
        |FROM SCRIPT $schemaName.KAFKA_CONSUMER WITH
        |BOOTSTRAP_SERVERS = 'kafka01:9092'
        |TOPIC_NAME        = '$topicName'
        |TABLE_NAME        = '${table.getFullyQualifiedName()}'
        |POLL_TIMEOUT_MS   = '300'
        |GROUP_ID          = 'exasol-kafka-udf-consumers'
    """.stripMargin

}

File: src/test/java/com/exasol/cloudetl/kafka/KafkaImportIT.java
package com.exasol.cloudetl.kafka;

import static com.exasol.matcher.ResultSetStructureMatcher.table;
import static com.exasol.matcher.TypeMatchMode.NO_JAVA_TYPE_CHECK;
import static org.hamcrest.MatcherAssert.assertThat;

import java.sql.*;
import java.util.*;

import org.apache.avro.*;
import org.apache.avro.generic.*;
import org.apache.kafka.common.serialization.*;
import org.hamcrest.Matcher;
import org.junit.jupiter.api.*;
import org.testcontainers.containers.Container.ExecResult;

import com.exasol.dbbuilder.dialects.Table;

import io.confluent.kafka.serializers.KafkaAvroSerializer;

class KafkaImportIT extends BaseKafkaDockerIntegrationTest {
    private static final String SCHEMA_NAME = "KAFKA_SCHEMA";
    private static final String TABLE_NAME = "KAFKA_TABLE";
    private String topicName;

    @BeforeAll
    void beforeAllKafkaImport() {
        installKafkaConnector(SCHEMA_NAME);
        updateExasolHostsFile();
    }

    @BeforeEach
    void beforeEachKafkaImport() {
        this.topicName = "kafka-topic-" + UUID.randomUUID();
        createTopic(this.topicName);
    }

    @AfterEach
    void afterEachKafkaImport() {
        executeStmt("DROP TABLE IF EXISTS " + getTableName());
        deleteTopic(this.topicName);
    }

    @Test
    void importsLongsAsTimestampValues() {
        TimeZone.setDefault(this.exasolContainer.getTimeZone());
        final Timestamp timestamp = new Timestamp(System.currentTimeMillis());

        new KafkaImportChecker(Map.of("COLUMN", "TIMESTAMP"))
                .withImportProperties(Map.of("RECORD_VALUE_FORMAT", "avro"))
                .withTopicValues(List.of(timestampRecord(0L), timestampRecord(1337L), timestampRecord(timestamp.getTime())),
                        timestampAvroSerializer())
                .assertResult(table().row(new Timestamp(0L), 0L, 0L).row(new Timestamp(1337L), 0L, 1L)
                        .row(timestamp, 0L, 2L).matches(NO_JAVA_TYPE_CHECK));
    }

    @Test
    void importsStringValues() {
        new KafkaImportChecker(Map.of("COLUMN", "VARCHAR(20)")).withTopicValues(List.of("a", "b", "c"),
                new StringSerializer()).assertResult(table().row("a", 0L, 0L).row("b", 0L, 1L).row("c", 0L, 2L)
                        .matches(NO_JAVA_TYPE_CHECK));
    }

    @Test
    void importWithConsumeAllOffsetParameterEmptyTopic() {
        new KafkaImportChecker(Map.of("COLUMN", "VARCHAR(20)"))
                .withImportProperties(Map.of("CONSUME_ALL_OFFSETS", "true"))
                .assertResult(table("VARCHAR", "BIGINT", "DECIMAL").matches());
    }

    @Test
    void importWithConsumeAllOffsetParameter() {
        final Map<String, String> properties = Map.of("CONSUME_ALL_OFFSETS", "true", "MAX_POLL_RECORDS", "2",
                "MIN_RECORDS_PER_RUN", "1", "MAX_RECORDS_PER_RUN", "2");
        new KafkaImportChecker(Map.of("COLUMN", "VARCHAR(20)")).withImportProperties(properties)
                .withTopicValues(List.of("one", "two", "three"), new StringSerializer())
                .assertResult(table().row("one", 0L, 0L).row("two", 0L, 1L).row("three", 0L, 2L)
                        .matches(NO_JAVA_TYPE_CHECK));
    }

    void updateExasolHostsFile() {
        final String brokerIpAddress = getContainerNetworkAddress(this.kafkaBrokerContainer);
        final String schemaRegistryIpAddress = getContainerNetworkAddress(this.schemaRegistryContainer);
        final List<String> commands = List.of("cp /etc/hosts /tmp/hosts", "sed --in-place '/kafka01/d' /tmp/hosts",
                "sed --in-place '/schema-registry/d' /tmp/hosts", "echo '" + brokerIpAddress + " kafka01' >> /tmp/hosts",
                "echo '" + schemaRegistryIpAddress + " schema-registry' >> /tmp/hosts", "cp /tmp/hosts /etc/hosts");
        for (final String command : commands) {
            try {
                final ExecResult exitCode = this.exasolContainer.execInContainer("/bin/sh", "-c", command);
                if (exitCode.getExitCode() != 0) {
                    throw new IllegalStateException("Could not run command '" + command + "' in Exasol container.");
                }
            } catch (final Exception exception) {
                throw new IllegalStateException("Could not run command '" + command + "' in Exasol container.", exception);
            }
        }
    }

    private String getTableName() {
        return "\"" + SCHEMA_NAME + "\".\"" + TABLE_NAME + "\"";
    }

    private GenericRecord timestampRecord(final long timestamp) {
        final Schema schema = new Schema.Parser().parse("{\"type\":\"record\",\"name\":\"TimestampRecord\","
                + "\"fields\":[{\"name\":\"timestamp\",\"type\":\"long\"}]}");
        return new GenericRecordBuilder(schema).set("timestamp", timestamp).build();
    }

    private Serializer<GenericRecord> timestampAvroSerializer() {
        final KafkaAvroSerializer serializer = new KafkaAvroSerializer();
        serializer.configure(Map.of("schema.registry.url", getExternalSchemaRegistryUrl()), false);
        @SuppressWarnings("unchecked")
        final Serializer<GenericRecord> typed = (Serializer<GenericRecord>) (Serializer<?>) serializer;
        return typed;
    }

    private String defaultImportStatement(final Table table) {
        return "IMPORT INTO " + table.getFullyQualifiedName() + "\n"
                + "FROM SCRIPT " + SCHEMA_NAME + ".KAFKA_CONSUMER WITH\n"
                + "BOOTSTRAP_SERVERS   = 'kafka01:9092'\n"
                + "SCHEMA_REGISTRY_URL = 'http://schema-registry:8081'\n"
                + "TOPIC_NAME          = '" + this.topicName + "'\n"
                + "TABLE_NAME          = '" + table.getFullyQualifiedName() + "'\n"
                + "POLL_TIMEOUT_MS     = '300'\n"
                + "GROUP_ID            = 'exasol-kafka-udf-consumers'\n";
    }

    private final class KafkaImportChecker {
        private final Table table;
        private final Map<String, String> columns;
        private Map<String, String> properties;

        private KafkaImportChecker(final Map<String, String> columns) {
            this.columns = columns;
            this.table = getTable();
        }

        KafkaImportChecker withImportProperties(final Map<String, String> properties) {
            this.properties = properties;
            return this;
        }

        <V> KafkaImportChecker withTopicValues(final List<V> values, final Serializer<V> serializer) {
            produceRecords(topicName, values, serializer);
            return this;
        }

        void assertResult(final Matcher<ResultSet> matcher) {
            executeStmt(getImportStatement());
            try (ResultSet resultSet = executeQuery("SELECT * FROM " + getTableName())) {
                assertThat(resultSet, matcher);
            } catch (final SQLException exception) {
                throw new IllegalStateException("Failed to close result set", exception);
            }
        }

        private Table getTable() {
            var tableBuilder = schema.createTableBuilder(TABLE_NAME);
            final Map<String, String> allColumns = new LinkedHashMap<>(this.columns);
            allColumns.put("KAFKA_PARTITION", "DECIMAL(18, 0)");
            allColumns.put("KAFKA_OFFSET", "DECIMAL(36, 0)");
            for (final Map.Entry<String, String> column : allColumns.entrySet()) {
                tableBuilder = tableBuilder.column(column.getKey(), column.getValue());
            }
            return tableBuilder.build();
        }

        private String getImportStatement() {
            final StringBuilder builder = new StringBuilder(defaultImportStatement(this.table));
            if (this.properties == null || !this.properties.containsKey("RECORD_KEY_FORMAT")) {
                builder.append("\nRECORD_KEY_FORMAT = 'string'");
            }
            if (this.properties == null || !this.properties.containsKey("RECORD_VALUE_FORMAT")) {
                builder.append("\nRECORD_VALUE_FORMAT = 'string'");
            }
            if (this.properties != null) {
                this.properties.forEach((key, value) -> builder.append("\n").append(key).append(" = '").append(value).append("'"));
            }
            builder.append(";");
            return builder.toString();
        }
    }
}
